src/core/SampleBuffer.cpp

```diff
@@ -957,7 +929,11 @@ void SampleBuffer::visualize( QPainter & _p, const QRect & _dr,
 	const int xb = _dr.x();
 	const int first = focus_on_range ? _from_frame : 0;
 	const int last = focus_on_range ? _to_frame : m_frames;
-	for( int frame = first; frame < last; frame += fpp )
+
+	auto start = m_reversed ? last : first;
+	auto offset = m_reversed ? first : last;
+	for ( int frame = start; m_reversed ? frame > offset : frame < offset;
+		  m_reversed ? frame -= fpp : frame += fpp )
```

Comment:
Don't mean to nitpick, everything else looks good to me, but I guess `end` would convey the idea better than `offset`. It feels like in this `for` loop the offset is actually `fpp`. Also, just a suggestion, maybe break the `for` in 3 lines for even better readability?

```suggestion
	auto end = m_reversed ? first : last;
	for( int frame = start;
		m_reversed ? frame > end : frame < end;
		m_reversed ? frame -= fpp : frame += fpp )
```

Reply:
Honestly, I would want this to be completely changed. I don't like the expression within the ternary operator, it's not easy to read at a glance. I think setting up variables first and having a single comparison would be easier to read.

What would happen if *somehow* `m_reversed` was changed mid-loop?

Reply:
The variables are set before, but the comparisons are different on each case (`m_reversed` being true or false, will result in a _higher than_ or _lower than_ comparison) so not sure how to replace that middle statement. Maybe that last one (`m_reversed ? frame -= fpp : frame += fpp`) could be replaced if an `offset` variable was used, something like: `auto offset = m_reversed ? -fpp : fpp;` and having `frame += offset;`

As for `m_reversed` being changed mid-loop, I guess we would need to protect the methods with mutexes if they aren't already, I'd expect they were (because drawing the sample while it changes could be an issue) but I haven't looked to be sure. Very nice point bringing it up! Right now I think the consequences of changing the sample while it's being draw is just that it might be draw wrongly, but with `m_reversed` being used as a conditional on the `for` statement we can end up with out of bound accesses and infinite loops, much worse.

Reply:
> As for m_reversed being changed mid-loop, I guess we would need to protect the methods with mutexes if they aren't already
If you don't compare during the loop we don't have to worry about that risk, and don't have to protect with mutex (as much) 😁 

> so not sure how to replace that middle statement
```cpp
auto startFrame = m_reversed ? last : first;
auto endFrame = m_reversed ? first : last;
auto nextFrame = m_reversed ? -1 : 1;
for (auto frame = startFrame; startFrame != lastFrame; frame += nextFrame)
{ ... }
```

Reply:
True, that should work! I think we still need mutexes (to prevent `m_reversed` changing between the variables being set) but in terms of readability it does looks better!

Reply:
> What would happen if _somehow_ `m_reversed` was changed mid-loop?

Well, according to testing, nothing! It just reverses and continues playing from the new positions.

Veratil seems to have provided a better alternative for this loop, so I'll go with it :)

Reply:
@russiankumar
Race conditions are hard to test, for the issue to be noticed you'd have to reverse the sample **exactly** when a thread was going through the drawing loop, so the issue might be there but the state that triggers it hasn't happened when you tested. If I'm not wrong, the race condition is still there even with the improvement that @Veratil suggested (if a thread changes `m_reversed` after setting `startFrame` and `endFrame` but before setting `nextFrame` for example). I'm not sure how multithreading works on LMMS, but I think there should be a mutex on every method that accesses (either reading or writing) members of an object. That's of course outside the scope of your PR, though having that `for` checking for `m_reversed` on each loop makes the issue more apparent (there are much more opportunities for the race condition to happen).

As for ways to not have the ternary on the `for` conditional, I can't think of any that wouldn't make it worse to read. The readability is not that bad IMHO. The biger problem is, as mentioned before, it kind of creates more room for the race condition to happen. I'm not sure what's the best approach here, should we add mutexes to the `SampleBuffer` class before merging this?

Reply:
From Dom in Discord:
`A solution for your particular problem would probably be to read m_reversed once and keep it in a local variable, so it keeps the same value throughout the function.`

Reply:
@Veratil You've probably seen it on Discord already, but copying it here: Dom suggested accessing the member variable just once, storing it on a local variable and using it instead to reduce the space where the race condition can happen. I think that could be a good call for being able to merge this PR before fixing the multi-threading issues. I'd add a "TODO" comment on the lines that store the member variable to a local one explaining this shouldn't be needed when the mutexes are added.

Reply:
Alright, that seems like a good idea, I'll implement that. Still have no idea what to do about the segfault: do I leave the ternary in loop or should I see if I can figure out a way to extract it and have it work?

Reply:
> I didn't look to see what `fpp` is

It means frames per pixel.

Here are some more alternatives:
```cpp
auto startFrame = m_reversed ? last : first;
auto count = (last-first) / fpp + 1;
auto step = m_reversed ? -fpp : fpp;
for (auto frame = startFrame, i = 0; i < count; frame += step, ++i)
{ /**/ }
for (size_t i = 0; i < count; ++i)
{ auto frame = startFrame + i * step; /**/ }
```

Reply:
Thanks! I'll test this out and see if this works, and I'll commit this if it does.